processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -251,6 +245,112 @@ public static boolean convertSegment(File toConvert, File converted, IndexSpec i
     public MMappedIndex mapDir(File inDir) throws IOException;
   }
 
+  public static void validateRowValues(
+      Rowboat rb1,
+      IndexableAdapter adapter1,
+      Rowboat rb2,
+      IndexableAdapter adapter2
+  )
+  {
+    final int[][] dims1 = rb1.getDims();
+    final int[][] dims2 = rb2.getDims();
+    if (dims1.length != dims2.length) {
+      throw new SegmentValidationException(
+          "Dim lengths not equal %s vs %s",
+          Arrays.deepToString(dims1),
+          Arrays.deepToString(dims2)
+      );
+    }
+    final Indexed<String> dim1Names = adapter1.getDimensionNames();
+    final Indexed<String> dim2Names = adapter2.getDimensionNames();
+    for (int i = 0; i < dims1.length; ++i) {
+      final int[] dim1Vals = dims1[i];
+      final int[] dim2Vals = dims2[i];
+      final String dim1Name = dim1Names.get(i);
+      final String dim2Name = dim2Names.get(i);
+      final Indexed<String> dim1ValNames = adapter1.getDimValueLookup(dim1Name);
+      final Indexed<String> dim2ValNames = adapter2.getDimValueLookup(dim2Name);
+
+      if (dim1Vals == null || dim2Vals == null) {
+        if (dim1Vals != dim2Vals) {
+          throw new SegmentValidationException(
+              "Expected nulls, found %s and %s",
+              Arrays.toString(dim1Vals),
+              Arrays.toString(dim2Vals)
+          );
+        } else {
+          continue;
+        }
+      }
+      if (dim1Vals.length != dim2Vals.length) {
+        // Might be OK if one of them has null. This occurs in IndexMakerTest
+        if (dim1Vals.length == 0 && dim2Vals.length == 1) {
+          final String dimValName = dim2ValNames.get(dim2Vals[0]);
+          if (dimValName == null) {
+            continue;
+          } else {
+            throw new SegmentValidationException(
+                "Dim [%s] value [%s] is not null",
+                dim2Name,
+                dimValName
+            );
+          }
+        } else if (dim2Vals.length == 0 && dim1Vals.length == 1) {
+          final String dimValName = dim1ValNames.get(dim1Vals[0]);
+          if (dimValName == null) {
+            continue;
+          } else {
+            throw new SegmentValidationException(
+                "Dim [%s] value [%s] is not null",
+                dim1Name,
+                dimValName
+            );
+          }
+        } else {
+          throw new SegmentValidationException(
+              "Dim [%s] value lengths not equal. Expected %d found %d",
+              dim1Name,
+              dims1.length,
+              dims2.length
+          );
+        }
+      }
+
+      for (int j = 0; j < Math.max(dim1Vals.length, dim2Vals.length); ++j) {
```

Comment:
if they were not same then it should have gone in the
if (dim1Vals.length != dim2Vals.length) {..} block above which would either result in exception or continue the for loop at line-266?